keps/sig-node/2535-ensure-secret-pulled-images/README.md

```diff
@@ -165,57 +201,289 @@ ensure all tenants have rights to the images that are already present on a host.
+        kubelet->>cache: DELETE _pullling_ ("img:tag", secret A)
+        kubelet->>A: continue starting pod
+    end
+    end
+```
 
-The max size of the cache will scale with the number of unique cache entries * the number of unique images that have not been garbage collected.
-It is not expected that this will be a significant number of bytes. Will be verified by actual use in Alpha and subsequent metrics in Beta.
+  2. The image is already present on the node and it's either allow-listed or a record
+of its pull with the given credentials is present in the cache:
+
+```mermaid
+sequenceDiagram
+    participant A as Pod A
+    participant kubelet
+    participant CRI
+    participant cache as persistent<br/>cache
+
+    A->>kubelet: request ("img:tag", secret A)
+    kubelet->>CRI: request ID("img:tag")
+    CRI->>kubelet: ID("img:tag")
+    rect rgb(253, 253, 190)
+    alt image present and set not to reverify
+        kubelet->>kubelet: don't reverify
+    else image present in cache
+        kubelet->>cache: GET _pulled_ (ID("img:tag"), secret A)
+        cache->>kubelet: <previous successful auth info>
+    end
+    end
+    kubelet->>A: continue starting pod
+```
 
-See `/var/lib/kubelet/image_manager_state` in [kubernetes/kubernetes#114847](https://github.com/kubernetes/kubernetes/pull/114847)
+##### Failure modes:
+
+We should always fail safe. If there's an error reading from the disk, the record
+MUST be assumed as non-existent. This might cause an addition request to pull an image.
+If it is successful, the original failing record should be overridden to reflect the
+new success.
+
+#### Cache Directory Structure
+
+A new subdirectory is introduced to the kubelet's directory: `image_manager`.
+The directory contains two further subdirectories: `pulling` and `pulled`. The former
+serves for kubelet to store the state about images that it intends to pull. The
+latter subdirectory should contain information about all the images that were
+successfully pulled by the kubelet.
+
+The figure below shows the Go structures that should appear in files in the
+subdirectories described above. These structures **are not part of the REST API**
+but belong to the `imagemanager.kubelet.config.k8s.io` configuration API group.
+They only ever appear as JSON structures in files outside the kubelet memory.
+
+`ImagePullIntent` is the structure that is present in files in the `image_manager/pulling`
+directory. The filename of each such file is a SHA-256 hash of the image spec prepended
+by `sha256~`.
+
+`ImagePulledRecord` is the structure that appears in files in the `image_manager/pulled`
+directory. The filename of each such file is a sanitized image ID provided by the
+container runtime. This KEP expects the image IDs provided to be in the form of `<hash-alg>:<hash-val>`,
+thus the sanitization consists of replacing the `<hash-alg>:` with `<hash-alg>~`.
+
+```go
+type ImagePullIntent struct {
```

Comment:
is it possible for multiple parallel pull attempts of the same image spec with different credentials to happen in the kubelet? if so, how are those represented in a single file?

Reply:
Talking through this, if intent-to-pull *only* records the image, and we handle a crash mid-pull by treating everything with an intent-to-pull record as needing reverification, we don't need *any* credential info in the intent file, and multiple parallel pulls of the same image can just ensure the intent file exists.

On kubelet start, we'd still want to look at dangling intent-to-pull records, and check if images for those exist on the node.
1. For dangling intent-to-pull images which *do* exist, create digest entries for those images indicating reverification was needed (this implies the ImagePulledRecord must be able to express "needs reverification", similar to what we'll do when we cannot read/decode an ImagePulledRecord), then clean up intent-to-pull
2. For dangling intent-to-pull entries which *don't* have an image, we could leave the intent entries in place (either permanently or for "a while" ...), which will indicate future attempts to use those exact same image specs should be reverified

Reply:
If `NeverVerifyPreloadedImages`, dangling intent-to-pull entries which do exist, will become a preloaded image. How to distinguish between preloaded and dangling if we don't have the annotation?
- In the case of https://github.com/kubernetes/enhancements/pull/3547/files#diff-77a289ef32ead94beffe8193738841bf801db831a6383ed7de7342ccb0e1241c, sometimes user may trigger a manual image pull using crictl to pull manually.

If NeverVerifyAllowlistedImages(if not in the list), we can just do a reverification.

Reply:
> dangling intent-to-pull entries which do exist, will become a preloaded image

that should not happen... a dangling intent-to-pull entry means the kubelet tried to pull and the image should be treated as a pulled-by-kubelet image with unknown credentials, which should require reverification on first use